core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -243,6 +243,46 @@ class SocketServerTest extends JUnitSuite {
   }
 
   @Test
+  def testZeroMaxConnectionsPerIp() {
+    val newProps = new Properties()
+    newProps.putAll(props)
+    newProps.setProperty(KafkaConfig.MaxConnectionsPerIpProp, "0")
+    newProps.setProperty(KafkaConfig.MaxConnectionsPerIpOverridesProp, "%s:%s".format("127.0.0.1", "5"))
+    val cfg = KafkaConfig.fromProps(newProps)
+    val mtr = new Metrics()
+    val cpr = new CredentialProvider(cfg.saslEnabledMechanisms)
+    val srv = new SocketServer(cfg, mtr, Time.SYSTEM, cpr)
+    srv.startup()
+    // make the maximum allowable number of connections
+    val conns = (0 until 5).map(_ => connect(srv))
+    // now try one more (should fail)
+    val conn = connect(srv)
+    conn.setSoTimeout(3000)
+    assertEquals(-1, conn.getInputStream().read())
+    conn.close()
+
+    // it should succeed after closing one connection
+    val address = conns.head.getInetAddress
+    conns.head.close()
+    TestUtils.waitUntilTrue(() => srv.connectionCount(address) < conns.length,
+      "Failed to decrement connection count after close")
+    val conn2 = connect(srv)
+    val serializedBytes = producerRequestBytes
+    sendRequest(conn2, serializedBytes)
+    val request = srv.requestChannel.receiveRequest(2000)
+    assertNotNull(request)
+
+    // now try to connect from the external facing interface, which should fail
+    val conn3 = connect(s = srv, localAddr = InetAddress.getLocalHost)
```

Comment:
Thanks, didn't know about NetworkInterface.
Refactoring the code to something like this:
```
    // now try to connect from the external facing interface, which should fail
    enumerationAsScalaIterator(NetworkInterface.getNetworkInterfaces).find(iface => !iface.isLoopback) match {
      case Some(nonLoopback) => {
        val conn3 = connect(s = srv, localAddr = InetAddress.getLocalHost)
        conn3.setSoTimeout(3000)
        assertEquals(-1, conn3.getInputStream().read())
        conn3.close()
      }
      case None => fail("Couldn't obtain the non-loopback network interface")
    }
```